FWCore/Services/plugins/Tracer.cc

```diff
@@ -525,104 +549,301 @@ void Tracer::preallocate(service::SystemBounds const& bounds) {
+
+    printPaths(out, pathsAndConsumes);
+    printEDModulesWithDependentModules(out, allModules, pathsAndConsumes);
+    printESModulesWithDependentModules(out, allESModules, pathsAndConsumes);
+    printEDModulesWithConsumes(out, allModules, pathsAndConsumes);
+    printESModulesWithConsumes(out, allESModules, pathsAndConsumes);
+  }
+}
+
+void Tracer::printPaths(LogAbsolute& out, PathsAndConsumesOfModulesBase const& pathsAndConsumes) const {
+  out << "paths:\n";
+  std::vector<std::string> const& paths = pathsAndConsumes.paths();
+  if (paths.empty()) {
+    out << "  --- there are no paths ---\n";
+  }
+  for (auto const& path : paths) {
+    out << "  " << path << "\n";
+  }
+  out << "end paths:\n";
+  std::vector<std::string> const& endpaths = pathsAndConsumes.endPaths();
+  if (endpaths.empty()) {
+    out << "  --- there are no endpaths ---\n";
+  }
+  for (auto const& endpath : endpaths) {
+    out << "  " << endpath << "\n";
+  }
+  for (unsigned int j = 0; j < paths.size(); ++j) {
+    std::vector<ModuleDescription const*> const& modulesOnPath = pathsAndConsumes.modulesOnPath(j);
+    out << "modules on path " << paths.at(j) << ":\n";
+    if (modulesOnPath.empty()) {
+      out << "  --- there are no modules on this path ---\n";
+    }
+    for (auto const& desc : modulesOnPath) {
+      out << "  " << desc->moduleLabel() << "\n";
+    }
+  }
+  for (unsigned int j = 0; j < endpaths.size(); ++j) {
+    std::vector<ModuleDescription const*> const& modulesOnEndPath = pathsAndConsumes.modulesOnEndPath(j);
+    out << "modules on endpath " << endpaths.at(j) << ":\n";
+    if (modulesOnEndPath.empty()) {
+      out << "  --- there are no modules on this endpath ---\n";
+    }
+    for (auto const& desc : modulesOnEndPath) {
+      out << "  " << desc->moduleLabel() << "\n";
     }
-    out << "end paths:\n";
-    std::vector<std::string> const& endpaths = pathsAndConsumes.endPaths();
-    for (auto const& endpath : endpaths) {
-      out << "  " << endpath << "\n";
+  }
+}
+
+void Tracer::printEDModulesWithDependentModules(LogAbsolute& out,
+                                                std::vector<ModuleDescription const*> const& allModules,
+                                                PathsAndConsumesOfModulesBase const& pathsAndConsumes) const {
+  out << "All modules and modules in the current process whose products they consume:\n"
+         "(This does not include modules from previous processes or the source)\n"
+         "(Exclusively considers consumed Event and EventSetup products, not Run, Lumi, or ProcessBlock products)\n";
+  if (allModules.empty()) {
+    out << "  --- there are no modules ---\n";
+  }
+  for (auto const& module : allModules) {
+    out << "  " << module->moduleName() << "/'" << module->moduleLabel() << "'\n";
+    unsigned int moduleID = module->id();
+    if (pathsAndConsumes.moduleDescription(moduleID) != module) {
+      throw cms::Exception("TestFailure")
```

Comment:
Why `TestFailure` exception category?

Reply:
Changed it to "Tracer". Just a cut and paste thing I missed. This is just a sanity check anyway. Hopefully the exception is not thrown under any circumstances. Thanks.